homeassistant/components/konnected.py

```diff
@@ -119,6 +128,35 @@ def device_discovered(service, info):
                             " but not specified in configuration.yaml",
                             discovered.device_id)
 
+    def manual_discovery(event):
+        """Init devices on the network with manually assigned addresses."""
+        specified = list(filter(  # list of devices where host and port is set
+            lambda dev: dev.get(CONF_HOST) and dev.get(CONF_PORT),
+            cfg.get(CONF_DEVICES)))
+
+        while specified:
+            for dev in specified:
+                _LOGGER.debug("Discovering Konnected device %s at %s:%s",
+                              dev.get(CONF_ID),
+                              dev.get(CONF_HOST),
+                              dev.get(CONF_PORT))
+                try:
+                    discovered = DiscoveredDevice(hass,
+                                                  dev.get(CONF_HOST),
+                                                  dev.get(CONF_PORT))
+                    if discovered.is_configured:
+                        discovered.setup()
+                        specified.remove(dev)
+                    else:
+                        _LOGGER.err("""
+                            Konnected device %s was manually configured, but
+                            its mac address is not found in configuration.yaml
+                        """, discovered.device_id)
+                except konnected.Client.ClientError as err:
+                    _LOGGER.error(err)
+                    time.sleep(10)  # try again in 10 seconds
```

Comment:
Sleep isn't allowed in our thread pool. We can convert the function to a coroutine, use async sleep, and make sure the rest of the function code is async safe.

Reply:
What if the rest of the function code isn't async safe? what are my alternatives? In this case, the underlying `konnected.Client` is a synchronous http library.
Is it possible to async schedule this function to run again in a specified amount of time?

Reply:
It's possible to schedule functions to run in the thread pool. So instead of scheduling this whole function on the thread pool, we could make it a coroutine and from that schedule only the requests on the thread pool.

The other alternative is, as you say, use the asynchronous scheduler to schedule the function to run again later. The helper to do this is called `call_later` in helper/event.py.

Reply:
Ok, I think I got what you are suggesting. Thanks for the help.
However now I notice that the discovery function is running in `MainThread` instead of a `SyncWorker`. That's ok because it's async now?

Reply:
actually, I don't think this is right. I'm not getting the syntax right to schedule `setup_device` (the synchronous part) on the thread pool. Is there an example I can look at?

Reply:
Nevermind -- I think I got it!